File: Swift/Graph/[백준 2606] 바이러스.swift
//
//  [백준 2606] 바이러스.swift
//  Algorithm
//
//  Created by 김민 on 4/8/24.
//
// 백준 2606 바이러스 https://www.acmicpc.net/problem/2606

/*
[📌 전략]
 그래프 탐색으로 풀 수 있는 문제. dfs/bfs 모두 가능하고 dfs 탐색으로 풀이하였다.
*/


let n = Int(readLine()!)!
var m = Int(readLine()!)!
var g = Array(repeating: [Int](), count: 102)

// 그래프 입력 받기 - 무방향 그래프이므로 두 번 append 필요
while m > 0 {
    let line = readLine()!.split(separator: " ").map { Int($0)! }
    let (a, b) = (line[0], line[1])
    g[a].append(b)
    g[b].append(a)
    m -= 1
}

var visited = Array(repeating: false, count: 102)
var ans = 0

// dfs
func dfs(_ k: Int) {
    visited[k] = true

    for nxt in g[k] { // k 정점의 인접 정점 중 방문하지 않은 정점 방문하기
        if visited[nxt] { continue }

        ans += 1
        dfs(nxt)
    }
}

dfs(1) // 1번 컴퓨터부터 시작
print(ans)

File: Python/그래프/[백준 2606] 바이러스.py
'''
[백준 2606] 바이러스
https://www.acmicpc.net/problem/2606

[📌 전략]
- 1번 노드를 시작으로 바이러스가 퍼져나가는 노드 개수 탐색 -> dfs 사용

[✅ 풀이]
1. 양방향 그래프 입력
2. dfs 탐색
    - 파라미터로 받은 cur 노드를 기준으로 이웃한 노드를 순회하며 dfs 함수 재귀 호출
3. dfs(1) 호출

[📝 기록]
LV: 실버3
풀이 시간: 15분
'''

import sys

input = sys.stdin.readline

n = int(input())
m = int(input())

g = [[] for _ in range(n+1)]
visited = [False] * (n+1)

for _ in range(m): 
    e, v = map(int, input().split())
    g[e].append(v)
    g[v].append(e)

ans = 0

def dfs(cur):
    global ans
    visited[cur] = True

    for nxt in g[cur]:
        if visited[nxt]: continue

        ans += 1
        dfs(nxt)

dfs(1)
print(ans)